src/Microsoft.Framework.TestHost/ProjectCommand.cs

```diff
@@ -3,46 +3,89 @@
+            var commandText = project.Commands[command];
+            var replacementArgs = CommandGrammar.Process(
+                commandText, 
+                (key) => GetVariable(environment, key),
+                preserveSurroundingQuotes: false)
+                .ToArray();
 
-            var environment = new ApplicationEnvironment(
-                project, 
-                oldEnvironment.RuntimeFramework, 
-                oldEnvironment.Configuration);
+            var entryPoint = replacementArgs[0];
+            args = replacementArgs.Skip(1).Concat(args).ToArray();
 
-            var newServices = new ServiceProvider(services);
-            newServices.Add(typeof(IApplicationEnvironment), environment);
+            if (string.IsNullOrEmpty(entryPoint) ||
+                string.Equals(entryPoint, "run", StringComparison.Ordinal))
+            {
+                entryPoint = project.EntryPoint ?? project.Name;
+            }
+
+            CallContextServiceLocator.Locator.ServiceProvider = services;
+            return await ExecuteMain(services, entryPoint, args);
+        }
+
+        private static string GetVariable(IApplicationEnvironment environment, string key)
+        {
+            if (string.Equals(key, "env:ApplicationBasePath", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.ApplicationBasePath;
+            }
+            if (string.Equals(key, "env:ApplicationName", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.ApplicationName;
+            }
+            if (string.Equals(key, "env:Version", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.Version;
+            }
+            if (string.Equals(key, "env:TargetFramework", StringComparison.OrdinalIgnoreCase))
+            {
+                return environment.RuntimeFramework.Identifier;
+            }
+
+            return Environment.GetEnvironmentVariable(key);
+        }
 
-            var applicationHost = new Microsoft.Framework.ApplicationHost.Program(
-                (IAssemblyLoaderContainer)services.GetService(typeof(IAssemblyLoaderContainer)),
-                environment,
-                newServices);
+        private static Task<int> ExecuteMain(IServiceProvider services, string entryPoint, string[] args)
+        {
+            Assembly assembly = null;
 
-            var options = (IRuntimeOptions)services.GetService(typeof(IRuntimeOptions));
-            if (options.CompilationServerPort.HasValue)
+            try
             {
-                args = new string[]
+                assembly = Assembly.Load(new AssemblyName(entryPoint));
+            }
+            catch (FileNotFoundException ex) when (new AssemblyName(ex.FileName).Name == entryPoint)
+            {
+                if (ex.InnerException is ICompilationException)
```

Comment:
Yeah this isn't really protecting anything anyway. The `entryPoint` is the test framework itself, not the test project.